Libc/gen/syslog.3
.\" Copyright (c) 1985, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)syslog.3	8.1 (Berkeley) 6/4/93
.\" $FreeBSD: src/lib/libc/gen/syslog.3,v 1.22 2001/10/01 16:08:51 ru Exp $
.\"
.Dd June 4, 1993
.Dt SYSLOG 3
.Os
.Sh NAME
.Nm closelog ,
.Nm openlog ,
.Nm setlogmask ,
.Nm syslog ,
.Nm vsyslog
.Nd control system log
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In syslog.h
.Ft void
.Fo closelog
.Fa void
.Fc
.Ft void
.Fo openlog
.Fa "const char *ident"
.Fa "int logopt"
.Fa "int facility"
.Fc
.Ft int
.Fo setlogmask
.Fa "int maskpri"
.Fc
.Ft void
.Fo syslog
.Fa "int priority"
.Fa "const char *message"
.Fa "..."
.Fc
.In syslog.h
.In stdarg.h
.Ft void
.Fo vsyslog
.Fa "int priority"
.Fa "const char *message"
.Fa "va_list args"
.Fc
.Sh DESCRIPTION
The
.Fn syslog
function
writes
.Fa message
to the system message logger.
The message is then written to the system console, log files,
logged-in users, or forwarded to other machines as appropriate.
(See
.Xr syslogd 8 . )
.Pp
The message is identical to a
.Xr printf 3
format string, except that
.Ql %m
is replaced by the current error
message.
(As denoted by the global variable
.Va errno ;
see
.Xr strerror 3 . )
A trailing newline is added if none is present.
.Pp
Newlines and other non-printable characters embedded in the message string are printed in an alternate format.
This prevents someone from using non-printable characters to construct misleading log messages in an output file.
Newlines are printed as "\\n",
tabs are printed as "\\t".
Other control characters are printed using a caret ("^") representation, for example "^M" for carriage return.
.Pp
The
.Fn vsyslog
function
is an alternate form in which the arguments have already been captured
using the variable-length argument facilities of
.Xr stdarg 3 .
.Pp
The message is tagged with
.Fa priority .
Priorities are encoded as a
.Fa facility
and a
.Em level .
The facility describes the part of the system
generating the message.
The level is selected from the following
.Em ordered
(high to low) list:
.Bl -tag -width LOG_AUTHPRIV
.It Dv LOG_EMERG
A panic condition.
This is normally broadcast to all users.
.It Dv LOG_ALERT
A condition that should be corrected immediately, such as a corrupted
system database.
.It Dv LOG_CRIT
Critical conditions, e.g., hard device errors.
.It Dv LOG_ERR
Errors.
.It Dv LOG_WARNING
Warning messages.
.It Dv LOG_NOTICE
Conditions that are not error conditions,
but should possibly be handled specially.
.It Dv LOG_INFO
Informational messages.
.It Dv LOG_DEBUG
Messages that contain information
normally of use only when debugging a program.
.El
.Pp
The
.Fn openlog
function
provides for more specialized processing of the messages sent
by
.Fn syslog
and
.Fn vsyslog .
The parameter
.Fa ident
is a string that will be prepended to every message.
The
.Fa logopt
argument
is a bit field specifying logging options, which is formed by
.Tn OR Ns 'ing
one or more of the following values:
.Bl -tag -width LOG_AUTHPRIV
.It Dv LOG_CONS
If
.Fn syslog
cannot pass the message to
.Xr syslogd 8
it will attempt to write the message to the console
.Pq Dq Pa /dev/console .
.It Dv LOG_NDELAY
Open the connection to
.Xr syslogd 8
immediately.
Normally the open is delayed until the first message is logged.
Useful for programs that need to manage the order in which file
descriptors are allocated.
.It Dv LOG_PERROR
Write the message to standard error output as well to the system log.
.It Dv LOG_PID
Log the process id with each message: useful for identifying
instantiations of daemons.
.El
.Pp
The
.Fa facility
parameter encodes a default facility to be assigned to all messages
that do not have an explicit facility encoded:
.Bl -tag -width LOG_AUTHPRIV
.It Dv LOG_AUTH
The authorization system:
.Xr login 1 ,
.Xr su 1 ,
.Xr getty 8 ,
etc.
.It Dv LOG_AUTHPRIV
The same as
.Dv LOG_AUTH ,
but logged to a file readable only by
selected individuals.
.It Dv LOG_CRON
The cron daemon:
.Xr cron 8 .
.It Dv LOG_DAEMON
System daemons, such as
.Xr routed 8 ,
that are not provided for explicitly by other facilities.
.It Dv LOG_FTP
The file transfer protocol daemons:
.Xr ftpd 8 ,
.Xr tftpd 8 .
.It Dv LOG_KERN
Messages generated by the kernel.
These cannot be generated by any user processes.
.It Dv LOG_LPR
The line printer spooling system:
.Xr lpr 1 ,
.Xr lpc 8 ,
.Xr lpd 8 ,
etc.
.It Dv LOG_MAIL
The mail system.
.It Dv LOG_NEWS
The network news system.
.It Dv LOG_SECURITY
Security subsystems, such as
.Xr ipfw 4 .
.It Dv LOG_SYSLOG
Messages generated internally by
.Xr syslogd 8 .
.It Dv LOG_USER
Messages generated by random user processes.
This is the default facility identifier if none is specified.
.It Dv LOG_UUCP
The uucp system.
.It Dv LOG_LOCAL0
Reserved for local use.
Similarly for
.Dv LOG_LOCAL1
through
.Dv LOG_LOCAL7 .
.El
.Pp
The
.Fn closelog
function
can be used to close the log file.
.Pp
The
.Fn setlogmask
function
sets the log priority mask to
.Fa maskpri
and returns the previous mask.
Calls to
.Fn syslog
with a priority not set in
.Fa maskpri
are rejected.
The mask for an individual priority
.Fa pri
is calculated by the macro
.Fn LOG_MASK pri ;
the mask for all priorities up to and including
.Fa toppri
is given by the macro
.Fn LOG_UPTO toppri ; .
The default allows all priorities to be logged.
.Sh RETURN VALUES
The routines
.Fn closelog ,
.Fn openlog ,
.Fn syslog ,
and
.Fn vsyslog
return no value.
.Pp
The routine
.Fn setlogmask
always returns the previous log mask level.
.Sh EXAMPLES
.Bd -literal -offset indent -compact
syslog(LOG_ALERT, "who: internal error 23");

openlog("ftpd", LOG_PID | LOG_NDELAY, LOG_FTP);

setlogmask(LOG_UPTO(LOG_ERR));

syslog(LOG_INFO, "Connection from host %d", CallingHost);

syslog(LOG_INFO|LOG_LOCAL2, "foobar error: %m");
.Ed
.Sh LEGACY SYNOPSIS
.Fd #include <syslog.h>
.Fd #include <stdarg.h>
.Pp
These include files are necessary for all functions.
.Sh SEE ALSO
.Xr logger 1 ,
.Xr compat 5 ,
.Xr syslogd 8
.Sh HISTORY
These
functions appeared in
.Bx 4.2 .
.Sh BUGS
Never pass a string with user-supplied data as a format without using
.Ql %s .
An attacker can put format specifiers in the string to mangle your stack,
leading to a possible security hole.
This holds true even if the string was built using a function like
.Fn snprintf ,
as the resulting string may still contain user-supplied conversion specifiers
for later interpolation by
.Fn syslog .
.Pp
Always use the proper secure idiom:
.Pp
.Dl syslog(LOG_ERR, "%s", string);


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


Libc/gen/syslog.c
/*
 * Copyright (c) 1999 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/socket.h>
#include <sys/syslog.h>
#include <sys/uio.h>
#include <sys/un.h>
#include <netdb.h>

#include <errno.h>
#include <fcntl.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <notify.h>
#include <asl.h>
#include <asl_private.h>

#ifdef __STDC__
#include <stdarg.h>
#else
#include <varargs.h>
#endif

#include <crt_externs.h>

#define	LOG_NO_NOTIFY	0x1000
#define	INTERNALLOG	LOG_ERR|LOG_CONS|LOG_PERROR|LOG_PID

#ifdef BUILDING_VARIANT
__private_extern__ int	_sl_LogFile;		/* fd for log */
__private_extern__ int	_sl_connected;		/* have done connect */
__private_extern__ int	_sl_LogStat;		/* status bits, set by openlog() */
__private_extern__ const char *_sl_LogTag;	/* string to tag the entry with */
__private_extern__ int	_sl_LogFacility;	/* default facility code */
__private_extern__ int	_sl_LogMask;		/* mask of priorities to be logged */
__private_extern__ int  _sl_NotifyToken;	/* for remote control of priority filter */
__private_extern__ int  _sl_NotifyMaster;	/* for remote control of priority filter */
#else /* !BUILDING_VARIANT */
__private_extern__ int	_sl_LogFile = -1;		/* fd for log */
__private_extern__ int	_sl_connected = 0;		/* have done connect */
__private_extern__ int	_sl_LogStat = 0;		/* status bits, set by openlog() */
__private_extern__ const char *_sl_LogTag = NULL;	/* string to tag the entry with */
__private_extern__ int	_sl_LogFacility = LOG_USER;	/* default facility code */
__private_extern__ int	_sl_LogMask = 0xff;		/* mask of priorities to be logged */
__private_extern__ int  _sl_NotifyToken = -1;	/* for remote control of max logged priority */
__private_extern__ int  _sl_NotifyMaster = -1;	/* for remote control of max logged priority */
#endif /* BUILDING_VARIANT */

__private_extern__ void _sl_init_notify();

#define NOTIFY_SYSTEM_MASTER "com.apple.system.syslog.master"
#define NOTIFY_PREFIX_SYSTEM "com.apple.system.syslog"
#define NOTIFY_PREFIX_USER "user.syslog"
#define NOTIFY_STATE_OFFSET 1000

/* notify SPI */
uint32_t notify_register_plain(const char *name, int *out_token);
const char *asl_syslog_faciliy_num_to_name(int);

/*
 * syslog, vsyslog --
 *	print message on log file; output is intended for syslogd(8).
 */
void
#ifdef __STDC__
syslog(int pri, const char *fmt, ...)
#else
syslog(pri, fmt, va_alist)
	int pri;
	char *fmt;
	va_dcl
#endif
{
	va_list ap;

#ifdef __STDC__
	va_start(ap, fmt);
#else
	va_start(ap);
#endif
	vsyslog(pri, fmt, ap);
	va_end(ap);
}

void
vsyslog(int pri, const char *fmt, va_list ap)
{
	int status, i, saved_errno, filter, rc_filter;
	time_t tick;
	pid_t pid;
	uint32_t elen, count;
	char *p, *str, *expanded, *err_str, hname[MAXHOSTNAMELEN+1];
	uint64_t cval;
	int fd, mask, level, facility;
	aslmsg msg;

	saved_errno = errno;

	/* Check for invalid bits. */
	if (pri & ~(LOG_PRIMASK | LOG_FACMASK))
	{
		syslog(INTERNALLOG, "syslog: unknown facility/priority: %x", pri);
		pri &= (LOG_PRIMASK | LOG_FACMASK);
	}

	level = LOG_PRI(pri);
	facility = pri & LOG_FACMASK;

	if (facility == 0) facility = _sl_LogFacility;

	/* Get remote-control priority filter */
	filter = _sl_LogMask;
	rc_filter = 0;

	_sl_init_notify();

	if (_sl_NotifyToken >= 0) 
	{
		if (notify_get_state(_sl_NotifyToken, &cval) == NOTIFY_STATUS_OK)
		{
			if (cval != 0)
			{
				filter = cval;
				rc_filter = 1;
			}
		}
	}

	if ((rc_filter == 0) && (_sl_NotifyMaster >= 0))
	{
		if (notify_get_state(_sl_NotifyMaster, &cval) == NOTIFY_STATUS_OK)
		{
			if (cval != 0)
			{
				filter = cval;
			}
		}
	}

	mask = LOG_MASK(level);
	if ((mask & filter) == 0) return;

	/* Build the message. */
	msg = asl_new(ASL_TYPE_MSG);

	if (_sl_LogTag == NULL) _sl_LogTag = *(*_NSGetArgv());
	if (_sl_LogTag != NULL) 
	{
		asl_set(msg, ASL_KEY_SENDER, _sl_LogTag);
	}

	str = (char *)asl_syslog_faciliy_num_to_name(facility);
	if (str != NULL) asl_set(msg, ASL_KEY_FACILITY, str);

	str = NULL;
	tick = time(NULL);
	asprintf(&str, "%lu", tick);
	if (str != NULL)
	{
		asl_set(msg, ASL_KEY_TIME, str);
		free(str);
	}

	str = NULL;
	pid = getpid();
	asprintf(&str, "%u", pid);
	if (str != NULL)
	{
		asl_set(msg, ASL_KEY_PID, str);
		free(str);
	}

	str = NULL;
	asprintf(&str, "%d", getuid());
	if (str != NULL)
	{
		asl_set(msg, ASL_KEY_UID, str);
		free(str);
	}

	str = NULL;
	asprintf(&str, "%u", getgid());
	if (str != NULL)
	{
		asl_set(msg, ASL_KEY_GID, str);
		free(str);
	}

	str = NULL;
	asprintf(&str, "%u", level);
	if (str != NULL)
	{
		asl_set(msg, ASL_KEY_LEVEL, str);
		free(str);
	}

	status = gethostname(hname, MAXHOSTNAMELEN);
	if (status < 0) asl_set(msg, ASL_KEY_HOST, "localhost");
	else asl_set(msg, ASL_KEY_HOST, hname);

	/* check for %m */
	count = 0;
	for (i = 0; fmt[i] != '\0'; i++)
	{
		if ((fmt[i] == '%') && (fmt[i+1] == 'm')) count++;
	}

	expanded = NULL;
	elen = 0;
	err_str = NULL;

	/* deal with malloc failures gracefully */
	if (count > 0)
	{
		err_str = strdup(strerror(saved_errno));
		if (err_str == NULL) count = 0;
		else
		{
			elen = strlen(err_str);
			expanded = malloc(i + (count * elen));
			if (expanded == NULL) count = 0;
		}
	}

	if (expanded == NULL) expanded = (char *)fmt;
	if (count > 0)
	{
		p = expanded;

		for (i = 0; fmt[i] != '\0'; i++)
		{
			if ((fmt[i] == '%') && (fmt[i+1] == 'm'))
			{
				memcpy(p, err_str, elen);
				p += elen;
				i++;
			}
			else
			{
				*p++ = fmt[i];
			}
		}

		*p = '\0';
	}

	if (err_str != NULL) free(err_str);

	vasprintf(&str, expanded, ap);
	if (count > 0) free(expanded);

	if (str != NULL)
	{
		asl_set(msg, ASL_KEY_MSG, str);

		/* Output to stderr if requested. */
		if (_sl_LogStat & LOG_PERROR)
		{
			p = NULL;
			if (_sl_LogStat & LOG_PID) asprintf(&p, "%s[%u]: %s", (_sl_LogTag == NULL) ? "???" : _sl_LogTag, pid, str);
			else asprintf(&p, "%s: %s", (_sl_LogTag == NULL) ? "???" : _sl_LogTag, str);

			if (p != NULL)
			{
				struct iovec iov[2];

				iov[0].iov_base = p;
				iov[0].iov_len = strlen(p);
				iov[1].iov_base = "\n";
				iov[1].iov_len = 1;
				writev(STDERR_FILENO, iov, 2);
				free(p);
			}
		}

		free(str);
	}

	/* Get connected, output the message to the local logger. */
	str = asl_format_message(msg, ASL_MSG_FMT_RAW, ASL_TIME_FMT_SEC, &count);
	if (str != NULL)
	{
		p = NULL;
		asprintf(&p, "%10u %s", count, str);
		free(str);

		if (p != NULL)
		{
			count += 12;
			if (_sl_connected == 0) openlog(_sl_LogTag, _sl_LogStat | LOG_NDELAY, 0);

			status = send(_sl_LogFile, p, count, 0);
			if (status< 0)
			{
				closelog();
				openlog(_sl_LogTag, _sl_LogStat | LOG_NDELAY, 0);
				status = send(_sl_LogFile, p, count, 0);
			}

			if (status >= 0)
			{
				free(p);
				asl_free(msg);
				return;
			}

			free(p);
		}
	}

	/*
	 * Output the message to the console.
	 */
	if (_sl_LogStat & LOG_CONS && (fd = open(_PATH_CONSOLE, O_WRONLY | O_NOCTTY | O_NONBLOCK)) >= 0)
	{
		count = 0;

		p = asl_format_message(msg, ASL_MSG_FMT_STD, ASL_TIME_FMT_LCL, &count);
		if (p != NULL)
		{
			struct iovec iov;

			/* count includes trailing nul */
			iov.iov_len = count - 1;
			iov.iov_base = p;
			writev(fd, &iov, 1);
	
			free(p);
		}

		close(fd);
	}

	asl_free(msg);
}

#ifndef BUILDING_VARIANT

static struct sockaddr_un SyslogAddr;	/* AF_UNIX address of local logger */

__private_extern__ void
_sl_init_notify()
{
	int status;
	char *notify_name;
	const char *prefix;

	if (_sl_LogStat & LOG_NO_NOTIFY)
	{
		_sl_NotifyMaster = -2;
		_sl_NotifyToken = -2;
		return;
	}

	if (_sl_NotifyMaster == -1)
	{
		status = notify_register_plain(NOTIFY_SYSTEM_MASTER, &_sl_NotifyMaster);
		if (status != NOTIFY_STATUS_OK) _sl_NotifyMaster = -2;
	}

	if (_sl_NotifyToken == -1)
	{
		_sl_NotifyToken = -2;

		notify_name = NULL;
		prefix = NOTIFY_PREFIX_USER;
		if (getuid() == 0) prefix = NOTIFY_PREFIX_SYSTEM;
		asprintf(&notify_name, "%s.%d", prefix, getpid());

		if (notify_name != NULL)
		{
			status = notify_register_plain(notify_name, &_sl_NotifyToken);
			free(notify_name);
			if (status != NOTIFY_STATUS_OK) _sl_NotifyToken = -2;
		}
	}
}

void
openlog(ident, logstat, logfac)
	const char *ident;
	int logstat, logfac;
{
	if (ident != NULL) _sl_LogTag = ident;

	_sl_LogStat = logstat;

	if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0) _sl_LogFacility = logfac;

	if (_sl_LogFile == -1)
	{
		SyslogAddr.sun_family = AF_UNIX;
		(void)strncpy(SyslogAddr.sun_path, _PATH_LOG, sizeof(SyslogAddr.sun_path));
		if (_sl_LogStat & LOG_NDELAY)
		{
			if ((_sl_LogFile = socket(AF_UNIX, SOCK_DGRAM, 0)) == -1) return;
			(void)fcntl(_sl_LogFile, F_SETFD, 1);
		}
	}

	if ((_sl_LogFile != -1) && (_sl_connected == 0))
	{
		if (connect(_sl_LogFile, (struct sockaddr *)&SyslogAddr, sizeof(SyslogAddr)) == -1)
		{
			(void)close(_sl_LogFile);
			_sl_LogFile = -1;
		}
		else
		{
			_sl_connected = 1;
		}
	}

	_sl_init_notify();
}

void
closelog()
{
	if (_sl_LogFile >= 0) {
		(void)close(_sl_LogFile);
		_sl_LogFile = -1;
	}
	_sl_connected = 0;
}

/* setlogmask -- set the log mask level */
int
setlogmask(pmask)
	int pmask;
{
	int omask;

	omask = _sl_LogMask;
	if (pmask != 0) _sl_LogMask = pmask;
	return (omask);
}

#endif /* !BUILDING_VARIANT */
